src/core/util/hugepage_mgr.cpp

```diff
@@ -0,0 +1,258 @@
+        } else {
+            m_hugepages[hugepage].nr_hugepages_free -= actual_size / hugepage;
+        }
+    }
+
+    // Statistics
+    m_stats.total_requested += actual_size;
+    if (ptr) {
+        m_hugepages[hugepage].nr_hugepages_allocated += actual_size / hugepage;
+        ++m_hugepages[hugepage].nr_allocations;
+        ++m_stats.allocations;
+        m_stats.total_allocated += actual_size;
+        m_stats.total_unused += actual_size - size;
+
+        size = actual_size;
+    } else {
+        ++m_stats.fails;
+    }
+    return ptr;
+}
+
+void hugepage_mgr::dealloc_hugepages(void *ptr, size_t size)
+{
+    int rc = munmap(ptr, size);
+    if (rc != 0) {
+        __log_info_dbg("munmap failed (errno=%d)", errno);
+    }
+    // We don't track released hugepages. Usually they're used for preallocation
+    // and not freed in runtime.
+}
+
+void hugepage_mgr::print_report(bool short_report /*=false*/)
+{
+    std::lock_guard<decltype(m_lock)> lock(m_lock);
+
+    const size_t ONE_MB = 1024U * 1024U;
+    std::vector<size_t> hugepages;
+
+    get_supported_hugepages(hugepages);
+    vlog_printf(VLOG_INFO, "Hugepages info:\n");
+    if (safe_mce_sys().hugepage_log2) {
+        vlog_printf(VLOG_INFO, "  User forced to use %lu kB hugepages (%s=%u).\n",
+                    (1LU << safe_mce_sys().hugepage_log2) / 1024U, SYS_VAR_HUGEPAGE_LOG2,
+                    safe_mce_sys().hugepage_log2);
+    }
+    for (size_t hugepage : hugepages) {
+        vlog_printf(VLOG_INFO, "  %zu kB : total=%u free=%u\n", hugepage / 1024U,
+                    get_total_hugepages(hugepage), get_free_hugepages(hugepage));
+    }
+
+    if (short_report) {
+        return;
+    }
+
+    vlog_printf(VLOG_INFO, "Hugepages statistics:\n");
+    for (size_t hugepage : hugepages) {
+        vlog_printf(VLOG_INFO, "  %zu kB : allocated_pages=%u allocations=%u\n", hugepage / 1024U,
+                    m_hugepages[hugepage].nr_hugepages_allocated,
+                    m_hugepages[hugepage].nr_allocations);
+    }
+    vlog_printf(VLOG_INFO, "  Total: allocations=%u fails=%u\n", m_stats.allocations,
+                m_stats.fails);
+    vlog_printf(VLOG_INFO, "  Total: allocated=%zuMB requested=%zuMB unused_space=%zuMB\n",
+                m_stats.total_allocated / ONE_MB, m_stats.total_requested / ONE_MB,
+                m_stats.total_unused / ONE_MB);
+}
+
+void hugepage_mgr::read_sysfs()
+{
+    DIR *dir = opendir("/sys/kernel/mm/hugepages/");
+    struct dirent *entry;
+
+    if (dir) {
+        while ((entry = readdir(dir))) {
+            if (strncmp(entry->d_name, "hugepages-", 10) == 0) {
+                std::string path = std::string("/sys/kernel/mm/hugepages/") + entry->d_name;
+                size_t key = atol(entry->d_name + 10) * 1024U;
+
+                m_hugepages[key].nr_hugepages_total = read_file_uint(path + "/nr_hugepages");
+                m_hugepages[key].nr_hugepages_free = read_file_uint(path + "/free_hugepages");
```

Comment:
How do we track free hugepages if other applications allocate hugepages?

Reply:
We don't. There is a race and this is the reason why allocation may fail even if we predicated a success. Even running update() method before each allocation wouldn't avoid possible race with other processes

Reply:
Maybe we shouldn't bother with free pages at all. but selecting the next preferable huge-page size. lets discuss it.